tests/index/crlf.c

```diff
@@ -24,6 +26,210 @@ void test_index_crlf__cleanup(void)
+	git_buf_puts(&expected_path_oid, ".obj");
+	git_buf_puts(&expected_path_fail, ".fail");
+
+	if (git_path_isfile(expected_path_oid.ptr)) {
+		if (add_bypath_ret)
+			goto done;
+
+		if (git_futils_readbuffer(&expected_contents, expected_path_oid.ptr) < 0)
+			goto done;
+
+		if (git_oid_fromstr(&oid, expected_contents.ptr))
+			goto done;
+		if (!git_oid_equal(&oid, &entry->id))
+			goto done;
+	}
+
+	if (git_path_isfile(expected_path_fail.ptr)) {
+		if (!add_bypath_ret)
+			goto done;
+
+		if (git_futils_readbuffer(&expected_contents, expected_path_fail.ptr) < 0)
+			goto done;
+
+		git_buf_rtrim(&expected_contents);
+
+		if (error.klass != GITERR_FILTER || strstr(error.message, expected_contents.ptr) == NULL)
+			goto done;
+
+		git__free(error.message);
+	}
+
+	failed = false;
+
+done:
+	if (failed) {
+		git_buf details = GIT_BUF_INIT;
+		git_buf_printf(&details, "filename=%s, safecrlf=%s, autocrlf=%s, attrs={%s}",
+			git_path_basename(actual_path->ptr), cd->safecrlf, cd->autocrlf, cd->attrs);
+		clar__fail(__FILE__, __LINE__,
+			"adding file did not worked as expected", details.ptr, 0);
+		git_buf_free(&details);
+	}
+
+	git__free(basename);
+	git_buf_free(&expected_contents);
+	git_buf_free(&expected_path_oid);
+	git_buf_free(&expected_path_fail);
+
+	return 0;
+}
+
+static void test_add_index(const char *safecrlf, const char *autocrlf, const char *attrs)
+{
+	git_buf attrbuf = GIT_BUF_INIT;
+	git_buf expected_dirname = GIT_BUF_INIT;
+	git_buf sandboxname = GIT_BUF_INIT;
+	git_buf reponame = GIT_BUF_INIT;
+	struct compare_data compare_data = { NULL, safecrlf, autocrlf, attrs };
+	const char *c;
+
+	git_buf_puts(&reponame, "crlf");
+
+	git_buf_puts(&sandboxname, "safecrlf_");
+	git_buf_puts(&sandboxname, safecrlf);
+
+	git_buf_puts(&sandboxname, ",autocrlf_");
+	git_buf_puts(&sandboxname, autocrlf);
+
+	if (*attrs) {
+		git_buf_puts(&sandboxname, ",");
+
+		for (c = attrs; *c; c++) {
+			if (*c == ' ')
+				git_buf_putc(&sandboxname, ',');
+			else if (*c == '=')
+				git_buf_putc(&sandboxname, '_');
+			else
+				git_buf_putc(&sandboxname, *c);
+		}
+
+		git_buf_printf(&attrbuf, "* %s\n", attrs);
+		cl_git_mkfile("crlf/.gitattributes", attrbuf.ptr);
+	}
+
+	cl_repo_set_string(g_repo, "core.safecrlf", safecrlf);
+	cl_repo_set_string(g_repo, "core.autocrlf", autocrlf);
+
+	cl_git_pass(git_index_clear(g_index));
+
+	git_buf_joinpath(&expected_dirname, "crlf_data", "checkin_results");
+	git_buf_joinpath(&expected_fixture, expected_dirname.ptr, sandboxname.ptr);
+	cl_fixture_sandbox(expected_fixture.ptr);
+
+	compare_data.dirname = sandboxname.ptr;
+	cl_git_pass(git_path_direach(&reponame, 0, add_and_check_file, &compare_data));
+
+	cl_fixture_cleanup(expected_fixture.ptr);
+	git_buf_free(&expected_fixture);
+
+	git_buf_free(&attrbuf);
+	git_buf_free(&expected_fixture);
+	git_buf_free(&expected_dirname);
+	git_buf_free(&sandboxname);
+	git_buf_free(&reponame);
+}
+
+static void set_up_workingdir(const char *name)
```

Comment:
This was inspired by https://github.com/libgit2/libgit2/pull/3199/files as we need the subdirectory for the test case results to compare with.